diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c
index d04c4ff244e1..eaabe2d1b3ee 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c
@@ -46,6 +46,7 @@
 #define WPA_OUI				"\x00\x50\xF2"	/* WPA OUI */
 #define WPA_OUI_TYPE			1
 #define RSN_OUI				"\x00\x0F\xAC"	/* RSN OUI */
+#define DPP_OUI             "\x50\x6F\x9A" /* DPP OUI */
 #define	WME_OUI_TYPE			2
 #define WPS_OUI_TYPE			4
 
@@ -1478,6 +1479,10 @@ static s32 brcmf_set_wpa_version(struct net_device *ndev,
 		brcmf_err("set wpa_auth failed (%d)\n", err);
 		return err;
 	}
+	/* get the value for debug */
+	val=0;
+	err = brcmf_fil_bsscfg_int_get(netdev_priv(ndev), "wpa_auth", &val);
+	brcmf_dbg(CONN, "DBG: GET wpa_auth = [0x%0x] \n", val);
 	sec = &profile->sec;
 	sec->wpa_versions = sme->crypto.wpa_versions;
 	return err;
@@ -1612,6 +1617,7 @@ brcmf_set_key_mgmt(struct net_device *ndev, struct cfg80211_connect_params *sme)
 		return 0;
 
 	err = brcmf_fil_bsscfg_int_get(netdev_priv(ndev), "wpa_auth", &val);
+	brcmf_dbg(CONN, "DBG SET KEY MGMT  to [0x%0x] \n", val);
 	if (err) {
 		brcmf_err("could not get wpa_auth (%d)\n", err);
 		return err;
@@ -1649,6 +1655,9 @@ brcmf_set_key_mgmt(struct net_device *ndev, struct cfg80211_connect_params *sme)
 		case WLAN_AKM_SUITE_PSK:
 			val = WPA2_AUTH_PSK;
 			break;
+		case WLAN_AKM_SUITE_DPP:
+			val=WPA2_AUTH_PSK; /*TODO : This is not the actual value for DPP */
+			break;
 		default:
 			brcmf_err("invalid cipher group (%d)\n",
 				  sme->crypto.cipher_group);
@@ -3861,7 +3870,8 @@ static s32 brcmf_configure_opensecurity(struct brcmf_if *ifp)
 static bool brcmf_valid_wpa_oui(u8 *oui, bool is_rsn_ie)
 {
 	if (is_rsn_ie)
-		return (memcmp(oui, RSN_OUI, TLV_OUI_LEN) == 0);
+		return ((memcmp(oui, RSN_OUI, TLV_OUI_LEN) == 0) ||
+            (memcmp(oui, DPP_OUI, TLV_OUI_LEN) == 0));
 
 	return (memcmp(oui, WPA_OUI, TLV_OUI_LEN) == 0);
 }
diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/p2p.c b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/p2p.c
index 456a1bf008b3..7106d8f8cdca 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/p2p.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/p2p.c
@@ -58,6 +58,7 @@
 #define P2P_PUB_AF_CATEGORY	0x04
 #define P2P_PUB_AF_ACTION	0x09
 #define P2P_AF_CATEGORY		0x7f
+#define P2P_DPP_OUI 		0x1A	/* DPP OUI */
 #define P2P_OUI			"\x50\x6F\x9A"	/* P2P OUI */
 #define P2P_OUI_LEN		3		/* P2P OUI length */
 
@@ -66,6 +67,9 @@
 #define DOT11_ACTION_CAT_OFF	0	/* category offset */
 #define DOT11_ACTION_ACT_OFF	1	/* action offset */
 
+/* Action Frame OUI */
+#define DPP_GAS_ACTION_OUI     11 /* OUI offset */
+
 #define P2P_AF_DWELL_TIME		200
 #define P2P_AF_MIN_DWELL_TIME		100
 #define P2P_AF_MED_DWELL_TIME		400
@@ -236,13 +240,39 @@ static bool brcmf_p2p_is_pub_action(void *frame, u32 frame_len)
 
 	if (pact_frm->category == P2P_PUB_AF_CATEGORY &&
 	    pact_frm->action == P2P_PUB_AF_ACTION &&
-	    pact_frm->oui_type == P2P_VER &&
+	    ((pact_frm->oui_type == P2P_VER) || (pact_frm->oui_type == P2P_DPP_OUI)) &&
 	    memcmp(pact_frm->oui, P2P_OUI, P2P_OUI_LEN) == 0)
 		return true;
 
 	return false;
 }
 
+/**
+ * brcmf_is_dpp_action() - true if dpp action type frame.
+ *
+ * @frame: action frame data.
+ * @frame_len: length of action frame data.
+ *
+ * Determine if action frame is DPP action type
+ */
+
+static bool brcmf_is_dpp_action (void *frame, u32 frame_len)
+{
+   struct brcmf_p2p_pub_act_frame *pact_frm;
+
+   if (frame == NULL)
+       return false;
+
+   pact_frm = (struct brcmf_p2p_pub_act_frame *)frame;
+   if (frame_len < sizeof(struct brcmf_p2p_pub_act_frame) - 1)
+       return false;
+
+   if (pact_frm->oui_type == P2P_DPP_OUI)
+       return true;
+
+   return false;
+}
+
 /**
  * brcmf_p2p_is_p2p_action() - true if p2p action type frame.
  *
@@ -1490,7 +1520,7 @@ int brcmf_p2p_notify_action_tx_complete(struct brcmf_if *ifp,
  * frame is transmitted.
  */
 static s32 brcmf_p2p_tx_action_frame(struct brcmf_p2p_info *p2p,
-				     struct brcmf_fil_af_params_le *af_params)
+				     struct brcmf_fil_af_params_le *af_params, bool is_dpp_frame)
 {
 	struct brcmf_cfg80211_vif *vif;
 	s32 err = 0;
@@ -1502,7 +1532,18 @@ static s32 brcmf_p2p_tx_action_frame(struct brcmf_p2p_info *p2p,
 	clear_bit(BRCMF_P2P_STATUS_ACTION_TX_COMPLETED, &p2p->status);
 	clear_bit(BRCMF_P2P_STATUS_ACTION_TX_NOACK, &p2p->status);
 
-	vif = p2p->bss_idx[P2PAPI_BSSCFG_DEVICE].vif;
+	vif = p2p->bss_idx[P2PAPI_BSSCFG_PRIMARY].vif;
+	/* if dpp frame or using as AP; send it to primary interface
+	 * Do not send to p2p interface when using as AP to avoid the crash */
+	if (is_dpp_frame || vif->wdev.iftype == NL80211_IFTYPE_AP) {
+		vif = p2p->bss_idx[P2PAPI_BSSCFG_PRIMARY].vif;
+		brcmf_dbg(INFO,"DBG: Using Wlan0 interface \n");
+	}
+	else {
+		vif = p2p->bss_idx[P2PAPI_BSSCFG_DEVICE].vif;
+		brcmf_dbg(INFO,"DBG: Using P2P interface \n");
+	}
+
 	err = brcmf_fil_bsscfg_data_set(vif->ifp, "actframe", af_params,
 					sizeof(*af_params));
 	if (err) {
@@ -1659,6 +1700,7 @@ bool brcmf_p2p_send_action_frame(struct brcmf_cfg80211_info *cfg,
 	struct afx_hdl *afx_hdl = &p2p->afx_hdl;
 	u16 action_frame_len;
 	bool ack = false;
+	bool is_dpp_oui=false;
 	u8 category;
 	u8 action;
 	s32 tx_retry;
@@ -1687,6 +1729,9 @@ bool brcmf_p2p_send_action_frame(struct brcmf_cfg80211_info *cfg,
 	config_af_params.search_channel = false;
 	config_af_params.extra_listen = false;
 
+	/* Identify DPP oui */
+	is_dpp_oui = brcmf_is_dpp_action(action_frame->data, action_frame_len);
+
 	if (brcmf_p2p_is_pub_action(action_frame->data, action_frame_len)) {
 		/* p2p public action frame process */
 		if (brcmf_p2p_pub_af_tx(cfg, af_params, &config_af_params)) {
@@ -1696,6 +1741,10 @@ bool brcmf_p2p_send_action_frame(struct brcmf_cfg80211_info *cfg,
 		}
 	} else if (brcmf_p2p_is_gas_action(action_frame->data,
 					   action_frame_len)) {
+		/* identify the DPP specific GAS frame */
+		if (action_frame->data[DPP_GAS_ACTION_OUI] == P2P_DPP_OUI)
+			is_dpp_oui=true;
+
 		/* service discovery process */
 		if (action == P2PSD_ACTION_ID_GAS_IREQ ||
 		    action == P2PSD_ACTION_ID_GAS_CREQ) {
@@ -1779,7 +1828,7 @@ bool brcmf_p2p_send_action_frame(struct brcmf_cfg80211_info *cfg,
 	tx_retry = 0;
 	while (!p2p->block_gon_req_tx &&
 	       (ack == false) && (tx_retry < P2P_AF_TX_MAX_RETRY)) {
-		ack = !brcmf_p2p_tx_action_frame(p2p, af_params);
+		ack = !brcmf_p2p_tx_action_frame(p2p, af_params, is_dpp_oui);
 		tx_retry++;
 	}
 	if (ack == false) {
diff --git a/include/linux/ieee80211.h b/include/linux/ieee80211.h
index c83478271c2e..e97f02659dd5 100644
--- a/include/linux/ieee80211.h
+++ b/include/linux/ieee80211.h
@@ -2860,7 +2860,8 @@ enum ieee80211_sa_query_action {
 #define WLAN_AKM_SUITE_FILS_SHA384		SUITE(0x000FAC, 15)
 #define WLAN_AKM_SUITE_FT_FILS_SHA256		SUITE(0x000FAC, 16)
 #define WLAN_AKM_SUITE_FT_FILS_SHA384		SUITE(0x000FAC, 17)
-
+/* DPP Specific AKM SUITE */
+#define WLAN_AKM_SUITE_DPP                  SUITE(0x506F9A,2)
 #define WLAN_MAX_KEY_LEN		32
 
 #define WLAN_PMK_NAME_LEN		16
